ksql-engine/src/main/java/io/confluent/ksql/analyzer/TableFunctionExpressionRewriter.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.analyzer;
+
+import io.confluent.ksql.engine.rewrite.ExpressionTreeRewriter.Context;
+import io.confluent.ksql.execution.expression.tree.ColumnReferenceExp;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.expression.tree.VisitParentExpressionVisitor;
+import io.confluent.ksql.function.FunctionRegistry;
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.ColumnRef;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class TableFunctionExpressionRewriter
+    extends VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> {
+
+  private int variableIndex = 0;
+  private final FunctionRegistry functionRegistry;
+
+  public TableFunctionExpressionRewriter(final FunctionRegistry functionRegistry) {
+    super(Optional.empty());
+    this.functionRegistry = functionRegistry;
+  }
+
+  @Override
+  public Optional<Expression> visitFunctionCall(
+      final FunctionCall node,
+      final Context<Void> context) {
+    final String functionName = node.getName().name();
+    if (functionRegistry.isTableFunction(functionName)) {
+      final ColumnName varName = ColumnName.generatedColumnName(variableIndex);
+      variableIndex++;
+      return Optional.of(
+          new ColumnReferenceExp(node.getLocation(), ColumnRef.of(Optional.empty(), varName)));
```

Comment:
**tl;dr** most of my concerns are resolved with the exposing of the select expressions in the plan nodes. I think this is a really neat abstraction that we could utilize for a lot beyond even the FlatMapNode and AggregateNode.

Talked offline about this with @purplefox - the conclusion we came to is that the most important part is that the Logical Plan is properly abstracted; we can change how we get there (rewriting in one way or another) and what we do from there (arranging the physical plan properly) afterwards.

I'm pretty convinced with Tim's new changes to expose the Select Expressions alongside the Logical schema and letting each node change that, we can build a Logical plan that makes a lot of sense to me:
- we address the issue of rewriting because we can have the FlatMapNode do the mapping from input fields to output fields
- we don't deviate that far from the SQL standard because I think the logical plan that we build will be identical either way
- we can basically remove the analysis step altogether for table functions

That will basically model it the way that we wanted (first change the schema of FOO to BAR, then do the projection). Whether we rewrite it to make that explicit (e.g. `SELECT projected FROM FOO, LATERAL(FOO.col1) as projected`) or we have that implicit (e.g. `SELECT UDTF(col1) FROM FOO` and resolve it in the logical plan) will result in the same topology.